sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala

```diff
@@ -896,22 +967,29 @@ private[joins] object LongHashedRelation {
 
     // Create a mapping of key -> rows
     var numFields = 0
+    val isOriginalInputEmpty: Boolean = !input.hasNext
+    var allNullColumnKeyExistsInOriginalInput: Boolean = false
     while (input.hasNext) {
       val unsafeRow = input.next().asInstanceOf[UnsafeRow]
       numFields = unsafeRow.numFields()
       val rowKey = keyGenerator(unsafeRow)
       if (!rowKey.isNullAt(0)) {
         val key = rowKey.getLong(0)
         map.append(key, unsafeRow)
+      } else if (!allNullColumnKeyExistsInOriginalInput) {
+        allNullColumnKeyExistsInOriginalInput = true
       }
     }
     map.optimize()
     new LongHashedRelation(numFields, map)
+      .setOriginalInputEmtpy(isOriginalInputEmpty)
+      .setAllNullColumnKeyExistsInOriginalInput(allNullColumnKeyExistsInOriginalInput)
+      .asInstanceOf[LongHashedRelation]
   }
 }
 
 /** The HashedRelationBroadcastMode requires that rows are broadcasted as a HashedRelation. */
-case class HashedRelationBroadcastMode(key: Seq[Expression])
+case class HashedRelationBroadcastMode(key: Seq[Expression], isNullAware: Boolean = false)
```

Comment:
Shall we have more hash relation implementations?
1. If the original input is empty, return a `EmptyHashedRelation`
2. If there is any null key, return `EmptyHashedRelationWithNullKeys`
3. return `UnsafeHashedRelation` or `LongHashRelation`.

I think it's a bit confusing to introduce the "original input" concept to hashed relation.

Reply:
let me try update to adopt EmptyHashedRelation and get rid of original input

Reply:
done